Kernel/Syscalls/posix_spawn.cpp

```diff
@@ -0,0 +1,134 @@
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Debug.h>
+#include <Kernel/Devices/BaseDevices.h>
+#include <Kernel/Devices/Device.h>
+#include <Kernel/Devices/Generic/NullDevice.h>
+#include <Kernel/Devices/TTY/TTY.h>
+#include <Kernel/FileSystem/Custody.h>
+#include <Kernel/Memory/Region.h>
+#include <Kernel/Tasks/PerformanceManager.h>
+#include <Kernel/Tasks/Process.h>
+#include <Kernel/Tasks/Scheduler.h>
+#include <Kernel/Tasks/ScopedProcessList.h>
+
+namespace Kernel {
+ErrorOr<FlatPtr> Process::sys$posix_spawn(Userspace<Syscall::SC_posix_spawn_params const*> user_params)
+{
+    VERIFY_PROCESS_BIG_LOCK_ACQUIRED(this);
+    TRY(require_promise(Pledge::exec));
+
+    auto params = TRY(copy_typed_from_user(user_params));
+
+    if (params.arguments.length > ARG_MAX || params.environment.length > ARG_MAX)
+        return E2BIG;
+    if (params.arguments.length == 0)
+        return EINVAL;
+
+    if (params.attr_data.ptr() != 0 || params.attr_data_size != 0 || params.serialized_file_actions_data.ptr() != 0 || params.serialized_file_actions_data_size != 0) {
+        // FIXME: Implement spawn attributes and spawn file actions handling.
+        return ENOTSUP;
+    }
+
+    auto path = TRY(get_syscall_path_argument(params.path));
+
+    auto copy_user_strings = [](auto const& list, auto& output) -> ErrorOr<void> {
+        if (!list.length)
+            return {};
+        Checked<size_t> size = sizeof(*list.strings);
+        size *= list.length;
+        if (size.has_overflow())
+            return EOVERFLOW;
+        Vector<Syscall::StringArgument, 32> strings;
+        TRY(strings.try_resize(list.length));
+        TRY(copy_from_user(strings.data(), list.strings, size.value()));
+        for (size_t i = 0; i < list.length; ++i) {
+            auto string = TRY(try_copy_kstring_from_user(strings[i]));
+            TRY(output.try_append(move(string)));
+        }
+        return {};
+    };
+
+    Vector<NonnullOwnPtr<KString>> arguments;
+    TRY(copy_user_strings(params.arguments, arguments));
+
+    Vector<NonnullOwnPtr<KString>> environment;
+    TRY(copy_user_strings(params.environment, environment));
+
+    auto credentials = this->credentials();
+    Process::Name name {};
+    Process::current().name().with([&name](auto& process_name) {
+        name.store_characters(process_name.representable_view());
+    });
+    auto [child, child_first_thread] = TRY(Process::create(name.representable_view(), credentials->uid(), credentials->gid(), pid(), m_is_kernel_process, vfs_root_context(), hostname_context(), current_directory(), nullptr, tty(), nullptr));
+
+    ArmedScopeGuard thread_finalizer_guard = [&child_first_thread]() {
+        SpinlockLocker lock(g_scheduler_lock);
+        child_first_thread->detach();
+        child_first_thread->set_state(Thread::State::Dying);
+    };
+
+    // Since we aren't copying parent's file descriptor table, set it up like create init process does
```

Comment:
This seems wrong, the `posix_spawn` documentation says that file descriptors should be kept except for ones with `FD_CLOEXEC` and `FD_CLOFORK` set (plus other restrictions that I don't think apply to us yet).

At the very least file descriptors 0, 1, 2 should be copied over, otherwise it's going to be confusing when interacting with spawned processes over stdio.